include/mlir/Dialect/SPIRV/SPIRVBase.td

```diff
@@ -227,6 +230,20 @@ class SPV_Optional<Type type> : Variadic<type>;
 // TODO(ravishankarm): From 1.4, this should also include Composite type.
 def SPV_SelectType : AnyTypeOf<[SPV_Scalar, SPV_Vector, SPV_AnyPtr]>;
 
+// The execution scope or memory scope of an operation. When used as a memory
+// scope, it specifies the distance of synchronization from the current
+// invocation. When used as an execution scope, it specifies the set of
+// executing invocations taking part in the operation.
+def SPV_Scope : IntOfWidths<[32]>;
+
+// Memory semantics define memory-order constraints, and on what storage classes
+// those constraints apply to. The memory order constrains the allowed orders in
+// which memory operations in this invocation can made visible to another
+// invocation. The storage classes specify to which subsets of memory these
+// constraints are to be applied. Storage classes not selected are not being
+// constrained.
+def SPV_MemorySemantics : IntOfWidths<[32]>;
```

Comment:
Not sure about this approach. Should really be using an enum to describe all the valid semantics.. The enum would be an I32Attr (see other enum definitions in SPIRVBase.td). You could then let the operation take a list of these attributes (i.e. and ArrayAttr) and then having the instructions take an array of these

Reply:
@MaheshRavishankar thanks for review, since that mask it would require to describe all possible combination of those bits  https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Memory_Semantics_-id- to describe a type.
Also, regarding to the spec the operands must be a ssa-id of integer:
OpControlBarrier %int_2, %int_1, %int_2376 
it would require to creating of constants  while serizalizing.
If it's ok, I will update the patch.
Thanks.

Reply:
Technically this memory semantics `<id>` can be generated from another instruction like `OpGenericPtrMemSemantics`. But that's for OpenCL. For Vulkan, which supports `Shader` capability, the spec [requires](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_shadervalidation_a_validation_rules_for_shader_a_href_capability_capabilities_a) "All `<id>` used for Scope and Memory Semantics must be of an OpConstant." So we should be fine here to embed this constant directly into the instruction as an attribute and provide better assembly sugar for it. So +1 to what Mehesh said. If later somebody would like to add support for OpenCL, we can then revisit this. :)

Generally we are missing a principled approach to these "bit enums". I'm thinking introducing something like `BitEnumAttr`. Similar to `Int32EnumAttr`, it describes each possible bit using a `BitEnumAttrCase`. Also similar to `EnumGen`, we can have a TableGen backend to generate conversion functions for them. It should generate conversion functions that take in a int32_t and returns a vector containing the names for each set bit, and vice versa. With that, it would be nice to have something like

```mlir
spv.ControlBarrier "Workgroup", "WorkGroup", "Acquire|UniformMemory"
```

or similar.

And this can be used for other SPIR-V enums like FunctionControl, etc. 

WDYT?